File: backend/app/api/data_factory.py
"""
造数工厂 API
- GET /api/data-factory/categories          工具分类与参数 Schema（前端表单 + MCP 展示共用）
- POST /api/data-factory/tools/{tool_name}  统一工具执行入口
- POST /api/data-factory/batch              批量执行
"""
import logging
from typing import Any, Dict, List

from fastapi import APIRouter, Body, Depends, HTTPException

from app.core.deps import get_current_user
from app.models.user import User
from app.services.data_tools import (
    SERVICE_REGISTRY,
    DataToolError,
    execute_tool,
    get_tool_meta,
)

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/data-factory", tags=["造数工厂"])


def _to_http_error(e: DataToolError) -> HTTPException:
    return HTTPException(
        status_code=e.http_status,
        detail={"code": e.code, "message": e.message, "detail": e.detail},
    )


@router.get("/categories")
def get_categories(
    current_user: User = Depends(get_current_user),
):
    """获取工具分类与工具清单（含参数 Schema）"""
    return {"categories": get_tool_meta(), "total": len(SERVICE_REGISTRY)}


@router.post("/tools/{tool_name}")
def run_tool(
    tool_name: str,
    params: Dict[str, Any] = Body(default={}),
    current_user: User = Depends(get_current_user),
):
    """统一工具执行入口"""
    try:
        result = execute_tool(tool_name, params)
    except DataToolError as e:
        raise _to_http_error(e)
    return {"tool": tool_name, "result": result}


@router.post("/batch")
def run_batch(
    items: List[Dict[str, Any]] = Body(...),
    current_user: User = Depends(get_current_user),
):
    """批量执行多个工具，单项失败不影响其余"""
    if not isinstance(items, list):
        raise HTTPException(400, detail={"code": "INVALID_PARAM", "message": "items 需为数组"})
    results = []
    for item in items:
        name = (item or {}).get("tool", "")
        params = (item or {}).get("params", {})
        try:
            result = execute_tool(name, params)
            results.append({"tool": name, "ok": True, "result": result})
        except DataToolError as e:
            results.append({"tool": name, "ok": False,
                            "error": {"code": e.code, "message": e.message, "detail": e.detail}})
    return results

File: aits_hub_web/src/api/dataFactory.ts
import request from './request'
import { dataPoolsApi } from './dataPools'

/** 工具参数 Schema（与后端 ToolParameter 同构） */
export interface ToolParamSchema {
  type: string
  properties: Record<string, any>
  required: string[]
}

/** 工具元信息 */
export interface DataToolMeta {
  name: string
  title: string
  description: string
  parameters: ToolParamSchema
  is_generator: boolean
}

/** 工具分类 */
export interface DataToolCategory {
  key: string
  title: string
  icon: string
  tools: DataToolMeta[]
}

/** 统一执行结果（生成类：{ count, result }；转换类：{ ... }） */
export interface ToolResult {
  count?: number
  result?: any
  [key: string]: any
}

export const dataFactoryApi = {
  /** 获取六类工具分类与参数 Schema */
  getCategories: () =>
    request.get<{ categories: DataToolCategory[]; total: number }>('/data-factory/categories'),

  /** 统一工具执行入口 */
  runTool: (toolName: string, params: Record<string, any>) =>
    request.post<{ tool: string; result: ToolResult }>(`/data-factory/tools/${toolName}`, params),

  /** 批量执行 */
  runBatch: (items: { tool: string; params: Record<string, any> }[]) =>
    request.post<{ tool: string; ok: boolean; result?: ToolResult; error?: any }[]>(
      '/data-factory/batch',
      items,
    ),

  /** 生成结果一键导入 Mock 数据池（生成 static 型数据池） */
  importToPool: (projectId: number, name: string, data: any[]) =>
    dataPoolsApi.create(projectId, { name, data, data_type: 'static' }),
}
